4060 consolidate translation strings with hook

fergie-nz opened the pull request:
<!-- IMPORTANT!
  - Every PR must reference an issue; this helps to explain the intent of the PR
 -->

Fixes #4060

# 👩🏻‍💻 What does this PR do?

Consolidates all translation json namespaces into 'common.json' for each language.

There were some duplications of translation keys in some cases. For some of these I have renamed some keys in components and in the common namespace so that we maintain unique translations.

For example we had the same key split across two namespaces:
"label.months" for "Months" and "months".
"error.no-items" for "no items" and "no items to display"
This has been split to remove ambiguity into:
```json
{
  "label.months": "months",
  "text.months": "Months",
  "error.no-items": "No items",
  "error.no-items-to-display": "No items to display."
}
```

In translated files, I have made similar splits into separate keys where the translation was obvious. Where there was any confusion, I have deleted both keys to remove the possibility of lingering false translations.
I've put all of this in its own commit in case we want to revert that, and keep one of the translation keys (9748dd9bbe054c35ffa0a4dd6bdc71de4bf66533). 
If we were to revert this commit and keep one of the keys, it would be somewhat arbitrary which we keep(either common or in alphabetical order of namespace where it occurs first). This is why I figured it is safer to remove so that translators can ensure we have accurate translations.
There were not many instances here of lost translations, so I thought this would be OKAY.

I've also added a pre-commit hook to sort out common.json files, and add these to the given commit. I'm unsure I've gone about this the right way so would love feedback here! :)

This PR also removes all calls for specific namespaces in useTranslation hook. 
I've left the hook unchanged however, in case we want to add client specific namespaces in future.

<!-- Explain the changes you made -->

<!-- why are the changes needed -->

<!-- Add a screenshot if there are UI changes  -->

## 💌 Any notes for the reviewer?

<!-- Do you have any specific questions for the reviewer? -->

<!-- Is there a high risk/complicated change they should focus on? -->

<!-- any general areas of the codebase touched? any side effects caused? -->

<!-- Anything half cooked but going to be finished off in a different PR? -->

# 🧪 Testing

I was wondering if there is a way to check translated files against weblate?
Very tricky to compare massive diffs especially re ordered.

# 📃 Documentation

- [x] **No documentation required**: no user facing changes or a bug fix which isn't a change in behaviour

fergie-nz commented:
This includes the ordering on commit bash script - but maybe will be moving this into a github action?

jmbrunskill commented on client/.husky/sort-json.sh:
```diff
@@ -0,0 +1,12 @@
+
+for FILE in "${TRANSLATION_FILES[@]}"; do
+  if [ -f "$FILE" ]; then
+    cat "$FILE" | jq --sort-keys >> output.tmp && mv output.tmp "$FILE"
```
Does this require `jq` installed?
I wonder if we should base it on the one from HSH instead?
https://github.com/msupply-foundation/health-supply-hub/tree/develop/frontend/.husky

roxy-dao replied:
Is this not being moved to actions?

jmbrunskill replied:
@fergie-nz  can you create an issue for creating an action for this?

fergie-nz replied:
https://github.com/msupply-foundation/open-msupply/issues/4971

jmbrunskill reviewed (changes requested):
Looks good Ferg, I think that pre-commit hook needs to change, but if we've agreed to move it to Actions, then we can just create the issue for that and remove from here.

Please make sure you create issues with all the keys to review in foreign languages?

fergie-nz commented:
Some more context: here are the keys I consolidated OR differentiated in english where duplicate keys occured:

For things like capitalisation or obvious translation, I followed the same rules in other languages

{
  "error.no-items": "No items",
  "error.no-items-alt": "No items to display.", // renamed to "error.no-items-to-display"
  "filename.asset-categories": "asset-categories",
  "filename.asset-categories-alt": "Asset Categories", // "Deleted alternate key"
  "label.target-demographic": "Target Demographic",
  "label.target-demographic-alt": "Target demographic", // "Removed non double capitalised"
  "heading.expiring-stock": "Expiring Stock",
  "heading.expiring-stock-alt": "Expiring stock", // "Removed non double capitalised"
  "button.add-item": "Add Item",
  "button.add-item-alt": "Add item", // "Removed non double capitalised"

  "button.return-lines": "Return selected lines",
  "button.return-lines-alt": "Process return of selected lines",
  // renamed to the following:
  "button.return-lines": "Process return of selected lines", // new default
  "button.replenishment-return-lines": "Return selected lines", // replenishment gets alt default

  "label.add-charges": "Add charges",
  "label.add-charges-alt": "Add charge", // renamed to "add-charge"
  "messages.no-related-documents": "No related documents",
  "messages.no-related-documents-alt": "No related documents.", // "Removed full stop version"

  "messages.confirm-delete-lines_one": "This will permanently remove 1 line from this prescription",
  "messages.confirm-delete-lines_one-alt": "This will permanently remove 1 line from this invoice",
  "messages.confirm-delete-lines_few": "This will permanently remove {{count}} lines from this prescription",
  "messages.confirm-delete-lines_few-alt": "This will permanently remove {{count}} lines from this invoice",
  "messages.confirm-delete-lines_many": "This will permanently remove {{count}} lines from this prescription",
  "messages.confirm-delete-lines_many-alt": "This will permanently remove {{count}} lines from this invoice",
  "messages.confirm-delete-lines_other": "This will permanently remove {{count}} lines from this prescription",
  "messages.confirm-delete-lines_other-alt": "This will permanently remove {{count}} lines from this invoice",
  // the following 4 added to differentiate between prescriptions:
  "messages.confirm-delete-invoice-lines_few": "This will permanently remove {{count}} lines from this invoice",
  "messages.confirm-delete-invoice-lines_many": "This will permanently remove {{count}} lines from this invoice",
  "messages.confirm-delete-invoice-lines_one": "This will permanently remove 1 line from this invoice",
  "messages.confirm-delete-invoice-lines_other": "This will permanently remove {{count}} lines from this invoice",

  "error.unable-to-scan": "Unable to scan QR code",
  "error.unable-to-scan-alt": "Unable to scan barcode: {{error}}",
  // differentiated
  "error.unable-to-scan": "Unable to scan QR code", // default
  "error.unable-to-scan-barcode": "Unable to scan barcode: {{error}}", // add this

  "label.group-by-item": "Group by Item",
  "label.group-by-item-alt": "Group by item", // deleted non caps

  "error.requisition-not-found": "Requisition not found",
  "error.requisition-not-found-alt": "Internal Order not found",
  "error.requisition-not-found": "Requisition not found", // default
  "error.order-not-found": "Internal Order not found", // add alternate key for order

  "label.amc": "AMC",
  "label.amc-alt": "AMC/AMD",
  "label.amc": "AMC", // default
  "label.amc/amd": "AMC/AMD", // add alternative

  "label.stock-on-hand": "SoH (Est. remaining)",
  "label.stock-on-hand-alt": "Stock on hand",
  "label.stock-on-hand-alt-alt": "Stock on Hand",
  // remove double caps and differentiate:
  "label.stock-on-hand-remaining": "SoH (Est. remaining)",
  "label.stock-on-hand": "Stock on hand",

  "messages.cant-return-shipment": "Cannot process return of lines until the status is 'Shipped'",
  "messages.cant-return-shipment-alt": "Cannot return lines until the status is 'Delivered'",
  // add new default for replenishment:
  "messages.cant-return-shipment": "Cannot process return of lines until the status is 'Shipped'",
  "messages.cant-return-shipment-replenishment": "Cannot return lines until the status is 'Delivered'",

  "messages.deleted-requisitions_few": "Deleted {{count}} requisitions",
  "messages.deleted-requisitions_few-alt": "Deleted {{count}} orders",
  "messages.deleted-requisitions_many": "Deleted {{count}} requisitions",
  "messages.deleted-requisitions_many-alt": "Deleted {{count}} orders",
  "messages.deleted-requisitions_other": "Deleted {{count}} requisitions",
  "messages.deleted-requisitions_other-alt": "Deleted {{count}} orders",
  // another split to order vs requisition:
  "messages.deleted-requisitions_one": "Deleted {{count}} requisition",
  "messages.deleted-requisitions_few": "Deleted {{count}} requisitions",
  "messages.deleted-requisitions_many": "Deleted {{count}} requisitions",
  "messages.deleted-requisitions_other": "Deleted {{count}} requisitions",
  // but we aren't even using requisitions keys anyway?
  "messages.deleted-orders_one": "Deleted {{count}} order",
  "messages.deleted-orders_few": "Deleted {{count}} orders",
  "messages.deleted-orders_many": "Deleted {{count}} orders",
  "messages.deleted-orders_other": "Deleted {{count}} orders"
}

fergie-nz commented:
Also some other options for tidying our translation files!

- When to capitalise vs when not to (different prefix key for caps vs non caps?)
- Should we have full stops after messages or warnings?
- Should we differentiate column header strings (where space might be an issue)
- Some niche cases also occur where the same value in english used across a number of parts of the app might need to be translated into different values in other languages. In this case we might need to add different keys for the same value. Feels like this could get confusing and/or easy to miss where devs will look up a key to match their value which might not point to the right value in translations. This has been fixed organically in some cases already where there are multiple duplicate keys spread across namespaces in other languages which don't exist in the english json files.

roxy-dao commented:
> Also some other options for tidying our translation files!
> 
> * When to capitalise vs when not to (different prefix key for caps vs non caps?)
> * Should we have full stops after messages or warnings?
> * Should we differentiate column header strings (where space might be an issue)
> * Some niche cases also occur where the same value in english used across a number of parts of the app might need to be translated into different values in other languages. In this case we might need to add different keys for the same value. Feels like this could get confusing and/or easy to miss where devs will look up a key to match their value which might not point to the right value in translations. This has been fixed organically in some cases already where there are multiple duplicate keys spread across namespaces in other languages which don't exist in the english json files.

Issue for the first one 😉 https://github.com/msupply-foundation/open-msupply/issues/3401

jmbrunskill reviewed (approved):
Thanks Fergus, Lets get this in before too many other changes come up!
